src/preprocess.rkt

```diff
@@ -7,40 +7,53 @@
 (provide find-preprocessing preprocess-pcontext remove-unnecessary-preprocessing)
 
 ;; See https://pavpanchekha.com/blog/symmetric-expressions.html
-(define (find-preprocessing expression context rules)
-  ;; Here `*` means a test identity that *may* be equal to `expression`, and
+(define (find-preprocessing initial specification context rules
+                            node-limit simplify?)
+  ;; Here `*` means a test identity that *may* be equal to `specification`, and
   ;; `~` means the simplest form of an expression.
-  (define variables (context-vars context))
-  (define variable-representations (context-var-reprs context))
   (define evens*
-    (for/list ([variable (in-list variables)]
-               [representation (in-list variable-representations)])
+    (for/list ([variable (in-list (context-vars context))]
+               [representation (in-list (context-var-reprs context))])
       ;; TODO: Handle case where neg isn't supported for this representation
       (define negate (get-parametric-operator 'neg representation))
-      (replace-vars (list (cons variable (list negate variable))) expression)))
-  (define pairs (combinations variables 2))
+      (replace-vars (list (cons variable (list negate variable))) specification)))
+  (define pairs (combinations (context-vars context) 2))
   (define swaps*
     (for/list ([pair (in-list pairs)])
       (match-define (list a b) pair)
-      (replace-vars (list (cons a b) (cons b a)) expression)))
-  (define query (make-egg-query (cons expression (append evens* swaps*)) rules))
-  (match-define (cons expression~ rest~) (map last (simplify-batch query)))
-  (define-values (evens~ swaps~) (split-at rest~ (length evens*)))
+      (replace-vars (list (cons a b) (cons b a)) specification)))
+  (define query
+    (let ([expressions* (cons specification (append evens* swaps*))])
+      (make-egg-query
+       (if simplify?
+           (cons initial expressions*)
+           expressions*)
+       rules)))
+  (define results (simplify-batch query))
+  (define simplified
+    (if simplify?
+        (remove-duplicates (first results))
+        (list initial)))
+  (match-define
+    (cons specification~ (app (curryr split-at (length evens*)) evens~ swaps~))
```

Comment:
This is very pretty! (Though I would add that in Herbie we typically use `xxx*` instead of `xxx~`)

Reply:
Thanks! It got a little more hairy when I got rid of the `if simplify?` checks, since we want each iteration with `initials`, but only the last one with all the other patterns.

I have a stash where I tried adding `egraph-find` back, and it makes `find-preprocessing` a few lines shorter since it gets rid of the need to monkey around with the list returned by `simplify-batch` to extract what we want out of it. I can commit that if you're not opposed to the idea.

Also, the tildes are a remainder from my janky naming scheme from earlier, but I got rid of them just now in favor of appending `-identities` to those variables instead.

Reply:
Uhh, not convinced on the simplify checks. Let’s do egraph-find in another PR.